tests/liealg/test_cartan_subalgebra.py

```diff
@@ -0,0 +1,414 @@
+adjvec_0 = np.random.random((3, 4))
+
+# To express the adjvec given in basis_0 in the standard Pauli basis, we need the inv sqrt Gram.
+# Compute the operator given by adjvec_0 via the new coefficient matrix adjvec_0 @ inv_sqrt_Gram_0
+inv_sqrt_Gram_0 = np.linalg.pinv(sqrtm(basis_coeffs_0 @ basis_coeffs_0.T))
+expected_0 = [qml.dot(c, basis_0).pauli_rep for c in adjvec_0 @ inv_sqrt_Gram_0]
+
+# Use the adjvec_0 for another 3 operators, now understood in the (orthonormal) basis
+# paulis_1_qubits. The expected output does not require a Gram matrix to be computed.
+expected_1 = [qml.dot(c, paulis_1_qubit).pauli_rep for c in adjvec_0]
+
+### Two-qubit test cases
+
+# Create a non-orthonormal basis on 2 qubits
+basis_coeffs_1 = np.random.random((16, 16))
+basis_1 = [qml.dot(c, paulis_2_qubits).pauli_rep for c in basis_coeffs_1]
+
+# Create an adjvec for 13 operators, which we understand to be in basis_1
+adjvec_1 = np.random.random((13, 16))
+
+# To express the adjvec given in basis_1 in the standard Pauli basis, we need the inv sqrt Gram.
+# Compute the operator given by adjvec_1 via the new coefficient matrix adjvec_1 @ inv_sqrt_Gram_1
+inv_sqrt_Gram_1 = np.linalg.pinv(sqrtm(basis_coeffs_1 @ basis_coeffs_1.T))
+expected_2 = [qml.dot(c, basis_1).pauli_rep for c in adjvec_1 @ inv_sqrt_Gram_1]
+
+# Use the adjvec_1 for another 13 operators, now understood in the (orthonormal) basis
+# paulis_2_qubits. The expected output does not require a Gram matrix to be computed.
+expected_3 = [qml.dot(c, paulis_2_qubits).pauli_rep for c in adjvec_1]
+
+# Collect all test cases. All operators are formatted as qml.pauli.PauliSentence
+ps_test_cases = [
+    (adjvec_0, basis_0, expected_0, False),  # Non-orthogonal basis
+    (adjvec_0, paulis_1_qubit, expected_1, True),  # Orthonormal basis
+    (adjvec_1, basis_1, expected_2, False),  # Non-orthogonal basis
+    (adjvec_1, paulis_2_qubits, expected_3, True),  # Orthonormal basis
+]
+
+# Translate test cases to qml.operation.Operations
+op_test_cases = [
+    (adj_vecs, [ps.operation() for ps in basis], [ps.operation() for ps in expected], is_ortho)
+    for adj_vecs, basis, expected, is_ortho in ps_test_cases
+]
+
+# Translate test cases to dense matrices
+dense_test_cases = [
+    (
+        adj_vecs,
+        [qml.matrix(ps, wire_order=[0, 1]) for ps in basis],
+        [qml.matrix(ps, wire_order=[0, 1]) for ps in expected],
+        is_ortho,
+    )
+    for adj_vecs, basis, expected, is_ortho in ps_test_cases
+]
+
+
+class TestAdjvecToOp:
+    """Test adjvec_to_op."""
+
+    def test_NotImplementedError(self):
+        """Test that NotImplementedError is raised"""
+        with pytest.raises(
+            NotImplementedError, match="At least one operator in the specified basis"
+        ):
+            _ = adjvec_to_op(
+                [
+                    [
+                        0.0,
+                    ]
+                ],
+                [qml.pauli.PauliWord({0: "X"})],
+            )
```

Comment:
This seems unnecessary?
```suggestion
            _ = adjvec_to_op([[0.0]], [qml.pauli.PauliWord({0: "X"})])
```

Reply:
black 🤷 😅